### DISCUSSION TITLE
Slower Citrix PVS boot with iPXE

### DISCUSSION THREAD

**Discussion by milo** [ipxe/ipxe#1310] [Q&A]
Hi,

we are using Citrix PVS to boot Windows 10 via legacy (not UEFI) PXE boot in our university classrooms. PVS technology in a short - you create Windows 10 image, upload it to PVS server, via `tftp-server-name` hint the PXE client to boot some Citrix PXE binary-image and after that, Windows 10 kernel is loaded and filesystem parts are downloaded on demand (or sort of). Such boot takes 1-2 minutes.

If we switch to iPXE - to provide dual boot with Ubuntu - PVS booting slows down to 9-10 minutes. Configuration for iPXE is simple:
```
:start
menu Boot menu - ${hostname}
...
...
:pvs
chain tftp://pvs.server.ip/ARDBP32.BIN  <<< Citrix PVS PXE binary
goto start

:ubuntu
kernel ...
initrd ...
boot
goto start
```

I don't know where to start dig. Any idea why it happens? Some kind of memory or network adapter initialization by iPXE and Citrix PXE binary is confused?

I compile iPXE from source, now we are using 3475f9162b84ce21327244ebce20ae29db6d7ac8.

I am grateful for any idea or debug procedure.

**Comment by mcb30** [comment]
Switch to using wimboot instead of PVS. Downloading and starting a WinPE image over HTTP using iPXE and wimboot should take about 10 seconds, so substantially faster than you are currently achieving.

**Reply by milo** [reply]
Good idea but hard to achive. Citrix PVS offers image snapshoting and versioning with great GUI and admins ACL. Moreover, we are running really high amount of software - AutoCAD, MS Office, BIM,Cinema, Matlab any many, many more (~300) - not sure about licensing and compatibility for WinPE. Thank you anyway.

**Reply by milo** [reply]
Oh, maybe I didn't get you. We are booting Windows 10 in ready-to use state. Full OS ready for work. Not some kind of cloning into local SSD or similar.

**Comment by NiKiZe** [comment]
I would do a packet capture for the 2 different ways to boot, and then try to spot differences.

It would also be interesting to see which files it grabs over TFTP to see if it can be done with wimboot instead.

One other data point we are missing is if you are using UNDI or native drivers (ipxe.pxe or undionly.kpxe) if you are not sure you can use `ifstat` to check which driver iPXE is using.

**Reply by milo** [reply]
To be sure, I've compiled from current master (3475f9162b84ce21327244ebce20ae29db6d7ac8).

I was using and now I'm using `bin/ipxe.pxe` and `show chip` reports `net0/chip:string = i219lm-2`.  I tried `imgfetch` and speed is great. 512MB in 4s (~1Gbps). So I tried to monitor ethernet port status on switch during boot.

Scenario 1) Booting PVS directly (`option bootfile-name "ARDBP32.BIN"`) - average speed is ~50 Mbps first 30 seconds and next 45 seconds ~340 Mbps.

Scenario 2) Booting PVS via iPXE (`chain tftp://.../ARDBP32.BIN`) - average speed is ~1.8 Mbps first 8 minutes and next 45 seconds ~330 Mbps.

Then I tried to use `bin/undionly.kpxe`. Made same `imgfetch` test, speed is still great ~1Gbps. But now booting Scenario 2 took same time as Scenatio 1. Verified 10 times. So probably problem solved but I don't know why.

Could someone please enlighten me?

**Reply by NiKiZe** [reply]
I have experienced scenarios where the native intel driver (i219lm-x) has been slower than UNDI.

UNDI is often preferred for pcbios chainloading, so a better option for your usecase. However, I would strongly avoid anything using TFTP, so if it was me, would check what it actually does, and migrate to something else.

**Reply by milo** [reply]
ARDBP32.BIN file is only 34kB in size. The file is provided by Citrix PVS environment, it contains IP adresses of Citrix PVS Stream Servers. When loaded, it makes DHCP request, randomly chooses PVS Stream Server and starts receiving image.

The last mentioned phase is slow in the Scenario 2 (1.8Mbps vs. 50Mbps). I don't know how iPXE affects that. Looks like network card is somehow initialized by iPXE and than ARDBP32.BIN cannot handle it. Or maybe not. I have no idea how this low-level things work.
